### ISSUE TITLE
Option to create tagged Union from Union() runtype ...?

### ISSUE THREAD

**Issue by breathe** [pelotom/runtypes#26]
It would be pretty nice if runtypes supported the ability to make a tagged Union from a runtype Union by adding an _type field not present in the original data with appropriate value ...  It would be useful when validating weird/legacy json structures that you don't control ...

Something like

```
const Asteroid = Record({
  location: Vector,
  mass: Number,
})

const Planet = Record({
  location: Vector,
  mass: Number,
  population: Number,
  habitable: Boolean,
})

const SpaceObject = TaggedUnion(Asteroid, Planet)

const asteroid = { ...}
const planet = { ... }

const asteroidOrPlanet = SpaceObject.check(asteroid)
console.log(asteroidOrPlanet._type) // "Asteroid"
console.log(SpaceObject.check(planet))._type // "Planet"
```

**Comment by pelotom** [comment]
This is an interesting idea, although I'm not sure it's in scope for the core library.

I don't think you'd want the result of `TaggedUnion` itself to be a `Runtype` however. An invariant that that I've consciously chosen to observe is that `check()` at runtime always just returns its argument or throws; it doesn't mutate anything or return new objects.

This is why the `Function` runtype can only check that something is a function, you can't make any checks having to do with the types of its arguments and return types, because

1. it cannot be done synchronously; all you can do is wrap the function in assertions which will be performed later, when the function is invoked
2. doing (1) requires returning a modified version of the function

Both of these are violations of the simple semantics of all other runtypes, so rather than make an ugly special case I created `Contract`s, which are not `Runtype`s but make use of them. I think what you're proposing falls into a similar category.

Then there is the question of whether such a construct belongs in this package or elsewhere. I'd consider including it here if you want to write a PR for it, but I think it could just as easily belong in a separate downstream microlibrary. In either case I probably won't have the spare cycles to implement it myself in the near future.

**Comment by breathe** [comment]
Hi — somehow I never saw this comment or forgot to reply ... — apologies.

I can take a look at implementing this with an approach like Contract.  I haven’t used the Contract interface for anything yet or dived into the details yet — does Contract produce a static type that is somehow derived from the type of the constraints — I would imagine that to serve the case of generic runtime constraints the produced type for the Contract would be semantically something like a nominal type and not actually representative of the type of the constraints?  

What I want to end up with is the same static type I’d get from the Union runtype plus a special field added by the “TaggedUnion” type transforming construct — that way I can use the field to statically map back to the static types of the union types despite the fact that the source data doesn’t itself contain the type discriminating member.  Given that I’m wanting to include the added _type in the static type — does this seem like it will be doable with an approach similar to the implementation of Contract?

(I haven’t dived into the details yet and just wanted to get a sense for whether it seemed like it could work).

**Comment by pelotom** [comment]
What you've outlined sounds reasonable in the abstract.

FYI, I just published a new release (`1.0.0`) which adds a `match` method to `Union`... you might take a look at that and see if it solves your use case, or at least it might give you some ideas. You'd use like, for example,

```ts
const SpaceObject = Union(Asteroid, Planet);

const isHabitable = SpaceObject.match([
  asteroid => false,
  planet => planet.habitable,
]);

if (isHabitable(spaceObject)) {
  // ...
}
```

**Comment by breathe** [comment]
Awesome — this looks very useful and might be just what I need!  I will play around with it.  :)

Thank you!!
